### PR TITLE
feat: check in performance test related code/utilities/scripts (1 of N)

### PR DESCRIPTION
**Pull request by michaelawyu**
### Description of your changes

This PR checks in the code we use in the triple 1K performance/scalability evaluation, which helps create 1K staged update runs and measure the completion latencies.

Related to #528 

I have:

- [x] Associated this change with a known KubeFleet Issue (Bug, Feature, etc).
- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

N/A

### Special notes for your reviewer

N/A

### CODE REVIEW

File: hack/perftest/1000stagedupdateruns/utils/poll.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+					continue
+				}
+
+				// Read the staged update run to check if it's completed.
+				var stagedUpdateRun placementv1beta1.ClusterStagedUpdateRun
+				if err := r.hubClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf(stagedUpdateRunNameFmt, resIdx)}, &stagedUpdateRun); err != nil {
+					fmt.Printf("long polling worker %d: failed to get staged update run run-%d: %v\n", workerIdx, resIdx, err)
+					// Requeue the run; no need to retry here.
+					r.toLongPollStagedUpdateRunsChan <- resIdx
+					time.Sleep(r.longPollingCoolDownPeriod)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/utils/poll.go:51]: N/A; benchmarking code -> the channel size has been set up so that it will not overflow. Though an idx might be retries for multiple times, it will only be recorded by the completion counter once.

File: hack/perftest/1000stagedupdateruns/utils/poll.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+					continue
+				}
+
+				// Check if the staged update run is completed.
+				runSucceededCond := meta.FindStatusCondition(stagedUpdateRun.Status.Conditions, string(placementv1beta1.StageUpdatingConditionSucceeded))
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/utils/poll.go]: Done.

File: hack/perftest/1000stagedupdateruns/utils/poll.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+						fmt.Printf("long polling worker %d exits\n", workerIdx)
+						return
+					}
+				case <-ctx.Done():
+					return
+				default:
+					if r.longPollingStagedUpdateRunsCount.Load() == r.completedStagedUpdateRunCount.Load() {
+						return
+					}
+					continue
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/utils/poll.go:42]: The spinning is by design, so that the worker can jump to the next item quickly. The condition part is fixed.

File: hack/perftest/1000stagedupdateruns/main.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+func main() {
+	ctx := context.Background()
+
+	// Read the arguments.
+	doCleanUp := false
+	cleanUpFlag := os.Getenv("CLEANUP")
+	if len(cleanUpFlag) != 0 {
+		doCleanUp = true
+	}
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/main.go:46]: N/A. Test-only code.

File: hack/perftest/1000stagedupdateruns/utils/cleanup.go [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+	r.cleanUpStrategy(ctx)
+	r.cleanUpRuns(ctx)
+}
+
+func (r *Runner) cleanUpStrategy(ctx context.Context) {
+	stagedUpdatedRunStrategy := &placementv1beta1.ClusterStagedUpdateStrategy{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: commonStagedUpdateRunStrategyName,
+		},
+	}
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/utils/cleanup.go]: Fixed.

File: hack/perftest/1000stagedupdateruns/README.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+This directory contains a utility program that creates 1K staged update runs that run concurrently and polls
+until their full completion. 
+
+The program is added for the purpose of testing the performance and scalability of KubeFleet.
+
+## Before you begin
+
+* Set up 1K placements using the utility program in `../1000placements` before running this utility program.
+* Make sure that all placements have been updated to use the `External` update strategy (via the staged update run APIs).
+* Make sure that all member clusters are labelled with `env=[canary|staging|production]` as appropriate so that KubeFleet
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/README.md]: Fixed.

File: hack/perftest/1000stagedupdateruns/README.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+## Running the utility program
+
+Run the commands below to run the utility program:
+
+```bash
+go run main.go
+```
+
+With the default setup, it might take 1-2 hours before all 1K staged update runs are fully completed. The programs
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/README.md:26]: Fixed.

File: hack/perftest/1000stagedupdateruns/utils/latency.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+		}
+	}()
+	wg.Wait()
+}
+
+func (r *Runner) TallyLatencyQuantiles() {
+	latencies := make([]float64, 0, len(r.stagedUpdateRunCompletionLatencyByRunName))
+	for _, latency := range r.stagedUpdateRunCompletionLatencyByRunName {
+		latencies = append(latencies, latency.Seconds())
+	}
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/utils/latency.go:41]: N/A. Test-only code; no need to catch all corner cases.

File: hack/perftest/1000stagedupdateruns/utils/cleanup.go [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+
+	errAfterReties := retry.OnError(r.retryOpsBackoff, func(err error) bool {
+		return err != nil && !errors.IsNotFound(err)
+	}, func() error {
+		return r.hubClient.Delete(ctx, stagedUpdatedRunStrategy)
+	})
+	if errAfterReties != nil && !errors.IsNotFound(errAfterReties) {
+		fmt.Printf("failed to delete staged update run strategy %s after retries: %v\n", commonStagedUpdateRunStrategyName, errAfterReties)
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/utils/cleanup.go]: Fixed.

File: hack/perftest/1000stagedupdateruns/example-strategy.yaml [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+    - name: canary
+      labelSelector:
+        matchLabels:
+          env: canary
+      maxConcurrency: 25%
+    - name: production
+      labelSelector:
+        matchLabels:
+          env: prod
+      maxConcurrency: 25%
```
[the remark this answers is not in the document]
Comment (michaelawyu) [reply hack/perftest/1000stagedupdateruns/example-strategy.yaml]: Fixed.

File: hack/perftest/1000stagedupdateruns/README.md [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+The program is added for the purpose of testing the performance and scalability of KubeFleet.
+
+## Before you begin
+
+* Set up 1K placements using the utility program in `../1000placements` before running this utility program.
+* Make sure that all placements have been updated to use the `External` update strategy (via the staged update run APIs).
+* Make sure that all member clusters are labelled with `env=[canary|staging|prod]` as appropriate so that KubeFleet
+can assign them to their respective stages.
+
+> If you have followed the instructions in `../../README.md` and have used the given scripts and utility programs
```
Comment (weng271190436) [inline hack/perftest/1000stagedupdateruns/README.md:15]: where is this README?
